pkg/controller/cachemanager/cachemanager.go

```diff
@@ -0,0 +1,417 @@
+		c.tracker.ForData(instance.GroupVersionKind()).CancelExpect(instance)
+		return nil
+	}
+
+	syncKey := syncutil.GetKeyForSyncMetrics(instance.GetNamespace(), instance.GetName())
+	_, err = c.opa.AddData(ctx, instance)
+	if err != nil {
+		c.syncMetricsCache.AddObject(
+			syncKey,
+			syncutil.Tags{
+				Kind:   instance.GetKind(),
+				Status: metrics.ErrorStatus,
+			},
+		)
+
+		return err
+	}
+
+	c.tracker.ForData(instance.GroupVersionKind()).Observe(instance)
+
+	c.syncMetricsCache.AddObject(syncKey, syncutil.Tags{
+		Kind:   instance.GetKind(),
+		Status: metrics.ActiveStatus,
+	})
+	c.syncMetricsCache.AddKind(instance.GetKind())
+
+	return err
+}
+
+func (c *CacheManager) RemoveObject(ctx context.Context, instance *unstructured.Unstructured) error {
+	// only perform work for watched gvks
+	if gvk := instance.GroupVersionKind(); !c.watchedSet.Contains(gvk) {
+		return nil
+	}
+
+	if _, err := c.opa.RemoveData(ctx, instance); err != nil {
+		return err
+	}
+
+	// only delete from metrics map if the data removal was succcesful
+	c.syncMetricsCache.DeleteObject(syncutil.GetKeyForSyncMetrics(instance.GetNamespace(), instance.GetName()))
+	c.tracker.ForData(instance.GroupVersionKind()).CancelExpect(instance)
+
+	return nil
+}
+
+func (c *CacheManager) wipeData(ctx context.Context) error {
+	if _, err := c.opa.RemoveData(ctx, target.WipeData()); err != nil {
+		return err
+	}
+
+	// reset sync cache before sending the metric
+	c.syncMetricsCache.ResetCache()
+	c.syncMetricsCache.ReportSync()
+
+	return nil
+}
+
+func (c *CacheManager) ReportSyncMetrics() {
+	c.syncMetricsCache.ReportSync()
+}
+
+func (c *CacheManager) listAndSyncDataForGVK(ctx context.Context, gvk schema.GroupVersionKind) error {
+	u := &unstructured.UnstructuredList{}
+	u.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   gvk.Group,
+		Version: gvk.Version,
+		Kind:    gvk.Kind + "List",
+	})
+
+	err := c.reader.List(ctx, u)
+	if err != nil {
+		return fmt.Errorf("replaying data for %+v: %w", gvk, err)
+	}
+
+	defer c.ReportSyncMetrics()
+
+	for i := range u.Items {
+		if err := c.AddObject(ctx, &u.Items[i]); err != nil {
+			return fmt.Errorf("adding data for %+v: %w", gvk, err)
+		}
+	}
+
+	return nil
+}
+
+func (c *CacheManager) manageCache(ctx context.Context) {
+	stopChan := make(chan bool, 1)
+	gvkErrdChan := make(chan schema.GroupVersionKind, 1024)
+	gvksFailingTolist := watch.NewSet()
+
+	gvksFailingToListReconciler := func(stopChan <-chan bool) {
+		for {
+			select {
+			case <-stopChan:
+				return
+			case gvk := <-gvkErrdChan:
+				gvksFailingTolist.Add(gvk)
+			}
+		}
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			close(stopChan)
+			close(gvkErrdChan)
+			return
+		case <-c.cacheManagementTicker.C:
+			c.mu.Lock()
+			c.makeUpdates(ctx)
+
+			// spin up new goroutines to relist if new gvks to relist are
+			// populated from makeUpdates.
+			if c.gvksToRelist.Size() != 0 {
+				// stop any goroutines that were relisting before
```

Comment:
Why are we starting to list again if previous lists have not completed?  That puts us in a situation where the time for listing is constrained to the length of the ticker.  

The purpose of the ticker isn't "Make sure all listing happens in this time", it's "this is the longest you should wait before checking if it is time to take action".

Reply:
Also, if we are worried about a single listing operation blocking the others due to it taking a long time, I believe that an interaction with the API server will timeout and thus can't hang indefinitely.

Reply:
Let me know if the concerns are not addressed now but IIRC we chatted offline about the need to interrupt the listing of gvks that need to be replayed in case the spec has once again changed. We want to make sure we have the most up to date version.